src/Features/CSharp/Portable/CodeMapper/CSharpMapCodeService.cs

```diff
@@ -0,0 +1,188 @@
+    {
+        var target = await GetMappingTargetAsync(document, focusLocations, cancellationToken).ConfigureAwait(false);
+
+        if ((await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false))?.Options is not CSharpParseOptions options)
+            return document;
+
+        using var _ = ArrayBuilder<TextChange>.GetInstance(out var changes);
+        foreach (var code in contents)
+        {
+            var mapHelper = await GetMapperHelperAsync(target, code, options, cancellationToken).ConfigureAwait(false);
+            if (mapHelper is not null)
+            {
+                var edits = await mapHelper.MapCodeAsync(cancellationToken).ConfigureAwait(false);
+                changes.AddRange(edits);
+            }
+        }
+
+        return await GetDocumentWithChangeAsync(document, changes.ToImmutable(), formatMappedCode, cancellationToken).ConfigureAwait(false);
+    }
+
+    private async Task<Document> GetDocumentWithChangeAsync(Document document, ImmutableArray<TextChange> textChanges, bool formatMappedCode, CancellationToken cancellationToken)
+    {
+        var oldText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+        if (formatMappedCode)
+        {
+            var cleanupOptions = await document.GetCodeCleanupOptionsAsync(
+                _globalOptions.GetCodeCleanupOptions(document.Project.Services, allowImportsInHiddenRegions: null, fallbackOptions: null),
+                cancellationToken).ConfigureAwait(false);
+            var formattingOptions = cleanupOptions.FormattingOptions;
+
+            var adjustedChanges = await AdjustTextChangesAsync(document, textChanges, formattingOptions.NewLine, cancellationToken).ConfigureAwait(false);
+            var spansToFormat = GetChangedSpansInNewText(adjustedChanges);
+
+            var newText = oldText.WithChanges(adjustedChanges);
+            var newDocument = document.WithText(newText);
+
+            return await Formatter.FormatAsync(newDocument, spansToFormat, formattingOptions, rules: null, cancellationToken).ConfigureAwait(false);
+        }
+        else
+        {
+            var newText = oldText.WithChanges(textChanges);
+            return document.WithText(newText);
+        }
+    }
+
+    private static async Task<DocumentSpan> GetMappingTargetAsync(Document document, ImmutableArray<DocumentSpan> focusLocations, CancellationToken cancellationToken)
+    {
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+        foreach (var focusLocation in focusLocations)
+        {
+            if (document.Id != focusLocation.Document.Id)
+                continue;
+
+            if (root.FullSpan.Contains(focusLocation.SourceSpan))
+                return focusLocation;
```

Comment:
Q: If we try to use the first `focusLocation` for mapping and it fails, would it be beneficial to then try to map with the rest of the `focusLocation`s?

Reply:
Yea, we could definitely do that, iterating over all the locations until one successful run